test_ask.py

```diff
@@ -3,24 +3,131 @@
+		self.assertEqual(expected, ask.route_path_to_func_name(test_value))
+
+	# maybe_place_space_before
+	def test_maybe_place_space_before(self):
+		token_val = '%'
+		test_value = '' 
+		expected = ' % '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	def test_maybe_place_space_before(self):
+		token_val = '%'
+		test_value = 'string' 
+		expected = 'string % '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	def test_maybe_place_space_before(self):
+		token_val = '%'
+		test_value = 'string\n' 
+		expected = 'string\n% '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	def test_maybe_place_space_before(self):
+		token_val = '%'
+		test_value = 'string\t' 
+		expected = 'string\t% '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	def test_maybe_place_space_before(self):
+		token_val = '%'
+		test_value = 'string(' 
+		expected = 'string(% '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	def test_maybe_place_space_before(self):
+		token_val = '%'
+		test_value = 'string ' 
+		expected = 'string % '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	def test_maybe_place_space_before(self):
+		token_val = '%'
+		test_value = 'string.' 
+		expected = 'string.% '
+		self.assertEqual(expected, ask.maybe_place_space_before(test_value, token_val))
+
+	#route_params
+	def test_route_params_empty_string(self):
+		test_value = ''
+		expected = ''
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_no_parameter(self):
+		test_value = 'no parameter given'
+		expected = ''
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_one_parameter(self):
+		test_value = '<param>'
+		expected = 'param'
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_two_parameters(self):
+		test_value = '<param1> and <param2>'
+		expected = 'param1, param2'
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_param_with_line_break(self):
+		test_value = 'string and <param\n>'
+		expected = 'param'
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_opened_not_closed(self):
+		test_value = '<param'
+		expected = ''
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_closed_not_opened(self):
+		test_value = 'param>'
+		expected = ''
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_empty_param(self):
+		test_value = '<>'
+		expected = ''
+		self.assertEqual(expected, ask.route_params(test_value))
+
+	def test_route_params_param_in_param(self):
+		test_value = '<par<param>ram>'
+		expected = 'param'
+		self.assertEqual(expected, ask.route_params(test_value))
+
+
+if __name__ == '__main__':
+    unittest.main()
```

Comment:
There was an indentation warning here when I opened this file in my text editor. Make sure that this line uses tabs and not spaces.

Reply:
I always use tabs so I don't know what triggered it. I've written it fresh just in case.